ivadomed/testing.py

```diff
@@ -0,0 +1,215 @@
+import os
+
+import numpy as np
+import torch
+import torch.backends.cudnn as cudnn
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from ivadomed import metrics as imed_metrics
+from ivadomed import utils as imed_utils
+from ivadomed.training import get_metadata
+from ivadomed.loader import utils as imed_loader_utils
+
+cudnn.benchmark = True
+
+
+def test(model_params, dataset_test, testing_params, log_directory, device, cuda_available=True,
+         metric_fns=None):
+    """Main command to test the network.
+
+    Args:
+        model_params (dict): Model's parameters.
+        dataset_test (imed_loader): Testing dataset
+        testing_params (dict):
+        log_directory (string):
+        device (torch.device):
+        cuda_available (Bool):
+        metric_fns (list):
+        debugging (Bool):
+    Returns:
+        dict: result metrics
+    """
+    # DATA LOADER
+    test_loader = DataLoader(dataset_test, batch_size=testing_params["batch_size"],
+                             shuffle=False, pin_memory=True,
+                             collate_fn=imed_loader_utils.imed_collate,
+                             num_workers=0)
+
+    # LOAD TRAIN MODEL
+    fname_model = os.path.join(log_directory, "best_model.pt")
+    print('\nLoading model: {}'.format(fname_model))
+    model = torch.load(fname_model, map_location=device)
+    if cuda_available:
+        model.cuda()
+    model.eval()
+
+    # CREATE OUTPUT FOLDER
+    path_3Dpred = os.path.join(log_directory, 'pred_masks')
+    if not os.path.isdir(path_3Dpred):
+        os.makedirs(path_3Dpred)
+
+    # METRIC MANAGER
+    metric_mgr = imed_metrics.MetricManager(metric_fns)
+
+    # UNCERTAINTY SETTINGS
+    if (testing_params['uncertainty']['epistemic'] or testing_params['uncertainty']['aleatoric']) and \
+            testing_params['uncertainty']['n_it'] > 0:
+        n_monteCarlo = testing_params['uncertainty']['n_it']
+        testing_params['uncertainty']['applied'] = True
+        print('\nComputing model uncertainty over {} iterations.'.format(n_monteCarlo))
+    else:
+        testing_params['uncertainty']['applied'] = False
+        n_monteCarlo = 1
+
+    for i_monteCarlo in range(n_monteCarlo):
+        preds_npy, gt_npy = run_inference(test_loader, model, model_params, testing_params, path_3Dpred,
+                                          cuda_available, i_monteCarlo, log_directory)
+        # TODO: adapt for multi-label
+        metric_mgr(preds_npy.squeeze(1), gt_npy.squeeze(1))
```

Comment:
metrics are already adapted to multi-label -> metric_mgr(preds_npy, gt_npy) why not this way? I modified it, but let me know if I'm missing something

Reply:
I had issues when playing with `test_inference` --> if I was not doing the squeeze at the end --> hausdorff was rejecting it.
Will investigate a bit further and be back!

Reply:
Ooooo that's simply that `test_inference` was using the 2D version of hausdorff --> all gravy with the 3D. Yayy